grafana-plugin/src/containers/ServiceNowConfigDrawer/ServiceNowStatusSection.tsx

```diff
@@ -0,0 +1,205 @@
+                  />
+                )}
+              />
+            </td>
+          </tr>
+
+          <tr>
+            <td>Resolved</td>
+            <td>
+              <Controller
+                name={'additional_settings.state_mapping.resolved'}
+                control={control}
+                render={({ field }) => (
+                  <Select
+                    {...field}
+                    value={field.value?.[1]}
+                    defaultValue={field.value?.[1]}
+                    menuShouldPortal
+                    className="select control"
+                    disabled={false}
+                    options={getAvailableStatusOptions(OnCallAGStatus.Resolved)}
+                    onChange={(option: SelectableValue) => {
+                      setValue(
+                        'additional_settings.state_mapping.resolved',
+                        option ? [option.label, option.value] : null
+                      );
+                      forceUpdate();
+                    }}
+                    {...selectCommonProps}
+                  />
+                )}
+              />
+            </td>
+          </tr>
+
+          <tr>
+            <td>Silenced</td>
+            <td>
+              <Controller
+                name={'additional_settings.state_mapping.silenced'}
+                control={control}
+                render={({ field }) => (
+                  <Select
+                    {...field}
+                    value={field.value?.[1]}
+                    menuShouldPortal
+                    className="select control"
+                    disabled={false}
+                    options={getAvailableStatusOptions(OnCallAGStatus.Silenced)}
+                    onChange={(option: SelectableValue) => {
+                      setValue(
+                        'additional_settings.state_mapping.silenced',
+                        option ? [option.label, option.value] : null
+                      );
+                      forceUpdate();
+                    }}
+                    {...selectCommonProps}
+                  />
+                )}
+              />
+            </td>
+          </tr>
+        </tbody>
+      </table>
+    </VerticalGroup>
+  );
+
+  function getAvailableStatusOptions(currentAction: OnCallAGStatus) {
+    const stateMapping = getValues()?.additional_settings?.state_mapping || {};
+    const keys = Object.keys(stateMapping);
+
+    // values are list of array-like values [label, id]
+    const values = keys
+      .map((k) => stateMapping[k])
+      .filter(Boolean)
+      .map((arr) => arr[1]);
+    const statusList = (alertReceiveChannelStore.serviceNowStatusList || []).map(([name, id]) => ({ id, name }));
+
+    return statusList
+      .filter((status) => values.indexOf(status.id) === -1 || stateMapping?.[currentAction]?.[0] === status.name)
+      .map((status) => ({
+        value: status.id,
+        label: status.name,
+      }));
+  }
```

Comment:
It's not easy to quickly understand how this code is mapping and filtering data. Maybe for things like that we could extract it out of component to helper and have simple unit test so that it's clear right away what's the input and output?

Reply:
I added `SelectableValue` so that it's more obvious it's mapping the data as required by `Select` component. I see what you mean by how would a helper help in this case?
- No other component really need an additional helper so that we justify abstracting logic to a helper and be able to even reuse that logic
- The filtering part is only a few lines, it just filters out the already selected values. Maybe a comment would make it more suggestive for you?

Reply:
It would be great if we start flattening our components by extracting things to helpers and store, leveraging pure functions more. In many places our components has grown to enormous sizes and most probably they also started with sth simple. But then, bit after bit, you add one small feature, another etc. The problem here is not a single bit, but the sum of all of them after some time :D And helpers defined outside components or custom hooks, even if used in a single places and not shared outside, are still a great way to make components cleaner.  Another benefit is that you don't create such function on every render.

So - it's a tiny thing in a single PR but there's a larger picture behind that that I'm trying to advocate for :D